server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java

```diff
@@ -83,27 +111,99 @@ public HealthIndicatorResult calculate(boolean includeDetails) {
+            + EnableAllocationDecider.Allocation.NONE.toString().toLowerCase(Locale.getDefault())
+            + "].",
+        null
+    );
+    public static final UserAction.Definition ACTION_ENABLE_TIERS = new UserAction.Definition(
+        "enable_data_tiers",
+        "Elasticsearch isn't allowed to allocate shards from these indices because the indices expect to be allocated to data tier nodes, "
+            + "but none were found in the cluster. Add nodes with data tier roles ("
+            + DataTier.DATA_HOT
+            + ", "
+            + DataTier.DATA_WARM
+            + ", "
+            + DataTier.DATA_COLD
+            + ", etc.) to the cluster.",
+        null
+    );
+    public static final UserAction.Definition ACTION_SHARD_LIMIT = new UserAction.Definition(
+        "increase_shard_limit",
+        "Elasticsearch isn't allowed to allocate some shards from these indices to any of the nodes in its data tier because each node in "
+            + "the tier has reached its shard limit. Increase the values for the ["
+            + ShardsLimitAllocationDecider.INDEX_TOTAL_SHARDS_PER_NODE_SETTING.getKey()
+            + "] index setting on each index or add more nodes to the target tiers.",
+        null
+    );
+    public static final UserAction.Definition ACTION_MIGRATE_TIERS = new UserAction.Definition(
+        "migrate_data_tiers",
+        "Elasticsearch isn't allowed to allocate some shards from these indices to any of the nodes in their data tiers because no nodes "
+            + "in the tiers are compatible with the allocation filters in the index settings. Remove the conflicting allocation filters "
+            + "from each index's settings or try migrating to data tiers using the data tier migration action "
+            + "[POST /_ilm/migrate_to_data_tiers].",
+        null
+    );
+    public static final UserAction.Definition ACTION_INCREASE_TIER_CAPACITY = new UserAction.Definition(
+        "increase_tier_capacity_for_allocations",
+        "Elasticsearch isn't allowed to allocate some shards from these indices to any of the nodes in their data tiers because there are "
+            + "not enough nodes in the tier to allocate each shard copy on a different node. Increase the number of nodes in this tier or "
+            + "decrease the number of replicas your indices are using.",
+        null
+
+    );
+
+    private class ShardAllocationCounts {
         private boolean available = true; // This will be true even if no replicas are expected, as long as none are unavailable
         private int unassigned = 0;
         private int unassigned_new = 0;
         private int unassigned_restarting = 0;
         private int initializing = 0;
         private int started = 0;
         private int relocating = 0;
-        private Set<String> indicesWithUnavailableShards = new HashSet<>();
+        private final Set<String> indicesWithUnavailableShards = new HashSet<>();
```

Comment:
This is random and I'm missing context, but would it be helpful to have this be a `TreeSet` so that the indices are always in a consistent order?

Reply:
It looks like this field is only pulled up in `getImpacts` and that method sorts the indices by priority, then name. Probably don't need to have them presorted?

That said, It might be good to have the indices sorted the same way in the user action output.

Reply:
Added in [422d39b](https://github.com/elastic/elasticsearch/pull/85071/commits/422d39b22890077784efaf2c90c6a64f95985bdc)